### DISCUSSION TITLE
WeakReferenceMessenger subscribers receive event multiple times

### DISCUSSION THREAD

**Discussion by needledrag** [dotnet/maui#14353] [Q&A]
When using messaging using WeakReferenceMessenger, on pages(view models) registered as transient I subscribe to a WeakReferenceMessenger message in the view model constructor.
Then navigating to the next page a message is sent to the registered recipients (previous page) passing a parameter when a certain action occurs.  This works great the first time this happens in the app.  
Close both pages and do the same process, the subscription then receives the message twice. Doing this again can even result in the message sent 3 times to the same subscription.

So either the transient objects (view models) aren't being removed from memory and the current page is receiving all the messages or the Messenger service isn't unsubscribing from the message when the page closes (which should result in it being disposed together with the message registration?)

I've tried checking to see if the WeakReferenceMessenger message IsRegistered before subscribing to it in the class but this always returns false, even though it seems there are now multiple subscriptions to it.

Anyone else experience a similar problem?
Thought of trying to dispose of the page myself but looking through messages and bug reports this doesn't work.

This is super frustrating and kind of negates using the Messaging in the CommunityToolkit.MVVM at all if I can't get around it.
Any help would be greatly appreciated.

**Comment by ygroenewald** [comment]
I am experiencing the same issue. Did you manage to find a solution or a workaround to this problem?

**Reply by needledrag** [reply]
I ended unsubscribing from the message when it's called and subscribing just before I knew I'd possibly need it.  Not optimal but it works for now.  Also added the check to see it I'd already registered it before reregistering. 

/
If (!WeakReferenceMessenger.Default.IsRegistered<StudioChangedMessage>(this))
                {
                    WeakReferenceMessenger.Default.Register<Messages.StudioChangedMessage>(this, (r, m) =>
                    {
                        MainThread.BeginInvokeOnMainThread(async () =>
                        {
                            await SetStudio();
                        });
                        WeakReferenceMessenger.Default.Unregister<StudioChangedMessage>(this);
                    });
                }

**Reply by needledrag** [reply]
Also unsubscribing when the page closes works but if you're using mvvm you might have to wire this up to your page events and do extra chexks, you have to check if you're subscribed first too.
Hopefully this is fixed soon

**Reply by ygroenewald** [reply]
Thanks for sharing!
